bigtable/tests/system.py

```diff
@@ -272,6 +272,85 @@ def test_delete_column_family(self):
         # Make sure we have successfully deleted it.
         self.assertEqual(temp_table.list_column_families(), {})
 
+    def test_retry(self):
+        import subprocess, os, stat, platform, ssl
+        from google.cloud.bigtable.client import Client
+        from google.cloud.bigtable.instance import Instance
+        from google.cloud.bigtable.table import Table
+
+        # import for urlopen based on version
+        try:
+            # python 3
+            from urllib.request import urlopen
+        except ImportError:
+            # python 2
+            from urllib2 import urlopen
+
+
+        TEST_SCRIPT = 'tests/retry_test_script.txt'
+        SERVER_NAME = 'retry_server'
+        SERVER_ZIP = SERVER_NAME + ".tar.gz"
+
+        def process_scan(table, range, ids):
+            range_chunks = range.split(",")
+            range_open = range_chunks[0].lstrip("[")
+            range_close = range_chunks[1].rstrip(")")
+            rows = table.read_rows(range_open, range_close)
+            rows.consume_all()
+
+        # Download server
+        MOCK_SERVER_URLS = {
+            'Linux': 'https://storage.googleapis.com/cloud-bigtable-test/retries/retry_server_linux.tar.gz',
```

Comment:
Where's the source for the mock server? Why is it a raw binary? This is a bit opaque. What if someone wants to update how that server works, how would they do that?

Can it be written in Python and part of this code base that's started up? Or can this be implemented as a feature in the [Bigtable emulator (written on Go)](https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/bigtable/bttest/inmem.go) which can be downloaded via `go get`?

Reply:
@mbrukman We actually control the CircleCI environment, so I'd be much happier if we embedded this mock server as part of the Dockerfile. WDYT? /cc @lukesneeringer

Reply:
I'm fine with that too, I just want to see the code behind this server in source control, so that it's available for inspection to understand what it does, or to debug future client changes, and possible for future developers to update/extend the server, if needed.

Reply:
The errors in the system tests seem to arise from this network call - the  "tar: Unexpected EOF in archive" failure suggests that the test is getting an empty archive from the network instead of the real server binary.  Futhermore, this tests passes locally for me, with my machine's internet connection.  Perhaps CI is not permitting this network call?

If that's true, it would be easier to just check in the server.  The server is currently written in Go, meaning we'd either have to check in multiple binaries (per platform) or take on a dependency on the go compiler in the system tests.  Alternatively, I could re-implement the server in Python, but that would be some more work.

The server was put at that web endpoint by @garye - Gary WDYT?

Reply:
It's certainly fine with me if we check in the server. Sounds like that's the easiest way forward right now.

Reply:
How about we add a wget/download step to the dockerfile for now that pulls the binary down from GCS, and I'll work on open-sourcing the tool itself? When it's out there I can adjust the dockerfile to do a `go get` / `go build` instead?

Reply:
@dhermes – would modifying [`.circleci/config.yml`](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/.circleci/config.yml) to add another `run` step be the right way of doing this?

Reply:
Yes, or just making a shell script that runs for Bigtable?

Reply:
@dhermes @calpeyser As a concrete proposal, what about incorporating this:
https://github.com/garye/google-cloud-python/commit/5979a5c070915d23e3841938f34364f4b1fae709

And then making the necessary adjustment to the test that uses the retry server?

Reply:
Seems like a fine workaround. I made comments on the commit but it's structurally the right idea.

Reply:
@dhermes Thanks I made some updates.
@calpeyser Can you update with something like the above commit when you get a chance?